testsuite/tests/contrib/lkt_semantic/array_type_expected/test.out

```diff
@@ -0,0 +1,314 @@
+     has type <StructDecl prelude: "Int">
+
+Expr <NumLit test.lkt:20:33-20:34>
+     has type <StructDecl prelude: "Int">
+
+Expr <NumLit test.lkt:20:36-20:37>
+     has type <StructDecl prelude: "Int">
+
+Expr <ArrayLiteral test.lkt:20:29-20:38>
+     has type <InstantiatedGenericType prelude: "Array[Int]">
+
+Expr <NumLit test.lkt:20:41-20:42>
+     has type <StructDecl prelude: "Int">
+
+Expr <NumLit test.lkt:20:44-20:45>
+     has type <StructDecl prelude: "Int">
+
+Expr <NumLit test.lkt:20:47-20:48>
+     has type <StructDecl prelude: "Int">
+
+Expr <ArrayLiteral test.lkt:20:40-20:49>
+     has type <InstantiatedGenericType prelude: "Array[Int]">
+
+Expr <ArrayLiteral test.lkt:20:28-20:50>
+     has type <InstantiatedGenericType prelude: "Array[Array[Int]]">
+
+Id   <RefId "Array" test.lkt:22:8-22:13>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Array" test.lkt:22:14-22:19>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Array" test.lkt:22:20-22:25>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Array" test.lkt:22:26-22:31>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Int" test.lkt:22:32-22:35>
+     references <StructDecl prelude: "Int">
+
+Expr <ArrayLiteral test.lkt:22:45-22:47>
+     has type <InstantiatedGenericType prelude: "Array[Int]">
+
+Expr <ArrayLiteral test.lkt:22:48-22:50>
+     has type <InstantiatedGenericType prelude: "Array[Int]">
+
+Expr <ArrayLiteral test.lkt:22:44-22:51>
+     has type <InstantiatedGenericType prelude: "Array[Array[Int]]">
+
+Expr <ArrayLiteral test.lkt:22:52-22:54>
+     has type <InstantiatedGenericType prelude: "Array[Array[Int]]">
+
+Expr <ArrayLiteral test.lkt:22:43-22:55>
+     has type <InstantiatedGenericType prelude: "Array[Array[Array[Int]]]">
+
+Expr <ArrayLiteral test.lkt:22:56-22:58>
+     has type <InstantiatedGenericType prelude: "Array[Array[Array[Int]]]">
+
+Expr <ArrayLiteral test.lkt:22:42-22:59>
+     has type <InstantiatedGenericType prelude: "Array[Array[Array[Array[Int]]]]">
+
+Id   <RefId "Array" test.lkt:24:17-24:22>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Array" test.lkt:24:23-24:28>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Int" test.lkt:24:29-24:32>
+     references <StructDecl prelude: "Int">
+
+test.lkt:24:38: error: Mismatched types: expected `Array[Int]`, got a number literal
+23 | @invalid val a: Array[Array[Int]] = [1]
+   |                                      ^ 
+
+Expr <ArrayLiteral test.lkt:24:37-24:40>
+     has type <InstantiatedGenericType prelude: "Array[Array[Int]]">
+
+Id   <RefId "Array" test.lkt:25:17-25:22>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Array" test.lkt:25:23-25:28>
+     references <GenericDecl prelude: "Array">
+
+Id   <RefId "Int" test.lkt:25:29-25:32>
+     references <StructDecl prelude: "Int">
+
+test.lkt:25:40: error: ambiguous type for expression
+24 | @invalid val a: Array[Array[Int]] = [[[1]]]
```

Comment:
This error message is surprising: I would have expected rather something like:
```
Mismatched types: expected `Int`, got `Array[Int]` instead
```
@raph-amiard what do you think?

Reply:
Shouldn't be something like:

```Mismatched types: expected `Array[Array[Array[Int]]]`, got `Array[Array[Int]]` instead```

?

Reply:
Actually, looks good to me as `[[[1]]]` type cannot be inferred. At the moment, we can only emit this kind of diagnostic in such cases.

Reply:
I still would like to have @raph-amiard’s opinion on this

Reply:
Oh I forgot to answer you:

We can’t infer the type of `[[[1]]]` without help from the context, so we can’t say that we got `Array[Array[Array[Int]]]`. What I would have thought we could do would have been:

* Try to match `Array[Array[Int]]` with `[[[1]]]`: both are arrays, so…
* Try to match `Array[Int]` with `[[1]]`: both are arrays, so…
* Try to match `Int` with `[1]`: one is an array and not the former, so we can emit `expected Int, got Array[Int] instead`